openwisp_controller/geo/approximate_location/tasks.py

```diff
@@ -0,0 +1,104 @@
+import logging
+
+from celery import shared_task
+from django.contrib.gis.geos import Point
+from swapper import load_model
+
+logger = logging.getLogger(__name__)
+
+
+@shared_task
+def manage_approximate_locations(device_pk, ip_address, add_existing=False):
+    """
+    Creates/updates approximate location for a device based on the latitude and
+    longitude or attaches an existing location if `add_existing` is True.
+    Existing location here means a location of another device whose last_ip matches
+    the given ip_address.
+
+    When `add_existing` is True:
+    - If the current device has no location, attach the existing one; if it does,
+    update it with the existing one only if it is approximate (fuzzy).
+
+    When `add_existing` is False:
+    - A new location is created if no location exists for current device, or
+    existing one is updated using coords from WHOIS record if it is approximate (fuzzy).
+    """
+    from openwisp_controller.config.whois.utils import send_whois_task_notification
+
+    Device = load_model("config", "Device")
+    Location = load_model("geo", "Location")
+    WHOISInfo = load_model("config", "WHOISInfo")
+    DeviceLocation = load_model("geo", "DeviceLocation")
+
+    device = Device.objects.get(pk=device_pk)
+    whois_obj = WHOISInfo.objects.filter(ip_address=ip_address).first()
+    device_location, _ = DeviceLocation.objects.select_related(
+        "location"
+    ).get_or_create(content_object_id=device_pk)
+    current_location = device_location.location
+
+    def _create_update_location():
+        coords = Point(whois_obj.longitude, whois_obj.latitude, srid=4326)
+        address = whois_obj.formatted_address
+        location_name = (
+            ",".join(address.split(",")[:2]) + f" (Estimated: {ip_address})"
```

Comment:
It should be similar with below name
```suggestion
            ",".join(address.split(",")[:2]) + f" (Estimated Location: {ip_address})"
```